lightstep/sdk/metric/internal/syncstate/sync.go

```diff
@@ -91,33 +98,77 @@ func (inst *Instrument) SnapshotAndProcess() {
 	inst.lock.Lock()
 	defer inst.lock.Unlock()
 
-	for key, rec := range inst.current {
-		if rec.conditionalSnapshotAndProcess(false) {
-			continue
+	for key, reclist := range inst.current {
+		// reclist is a list of records for this fingerprint.
+		var head *record
+		var tail *record
+
+		// Scan reclist and modify the list. We're holding the
+		// lock giving exclusive access to the head-of-list
+		// and each next field, so the process here builds a new
+		// linked list after filtering records that are no longer
+		// in use.
+		for rec := reclist; rec != nil; rec = rec.next {
+			if inst.singleSnapshotAndProcess(key, rec) {
+				if head == nil {
+					// The first time a record will be kept,
+					// it becomes the head and tail.
+					head = rec
+					tail = rec
+				} else {
+					// Subsequently, update the
+					// tail of the list.
+					tail.next = rec
```

Comment:
Doesn't this mean that the tail record is pointing to itself? Should this be rec.next?

Reply:
Or does the `tail.next = nil` on line 134 prevent this?

Reply:
Yes. I added a note to clarify that the list becomes temporarily invalid.